### ISSUE TITLE
Add removed_blocks property to layout class (and use it for the Hero block)

### ISSUE THREAD

**Issue by indigoxela** [backdrop/backdrop-issues#4209]
**Description of the need**

Currently there's no way for block modules to catch the removal of their blocks from layouts.
This also affects core modules.

**Proposed solution**

After some experiments it turned out that hook_layout_presave() fits for this use case, but layouts need an additional property to record block removal.

My use case:

A custom module handles file usage via file_usage_add/file_usage_delete in blocks containing a file. If the block gets updated, I check if the file has changed and add/remove related file usage (mark as permanent/temporary). That works fine so far.

I couldn't find any solution for the situation when _the block gets removed from a layout_.

Without that functionality, the file will stay in status "permanent", even though the block it was used in doesn't exist anymore.

By the way, the hero block is also affected and would benefit from a hook like that.

---
PR https://github.com/backdrop/backdrop/pull/3019

**Comment by indigoxela** [comment]
I've solved my problem with the mentioned contrib module in a different way.

Nevertheless a _hook_block_remove_ could be an enhancement. Are there any opinions?

**Comment by indigoxela** [comment]
So... here's something more "real" to discuss about: https://github.com/backdrop/backdrop/pull/3017

This also addresses _some_ of the concerns mentioned in https://github.com/backdrop/backdrop-issues/issues/2137

What it does:

The hook strikes right before the block data get deleted from layout config (happens _after_ the "Save layout" button is clicked).

Modules can implement it to, for instance, take care of file_usage_delete() or do some necessary database queries.

As an example the PR uses the hook for hero blocks, to remove file usage. BlockHero only exists inside layouts, so there's no other chance to take care of file usage after the block has been deleted.

Please note: the PR is incomplete. At least the api documentation is missing.

Feedback is welcome.

**Comment by docwilmot** [comment]
> happens after the "Save layout" button is clicked

I dont think that is the case. This method only removes the block from the config, and is called when you click the remove button in the layout drag drop page. Thats before you click Save.

**Comment by indigoxela** [comment]
> This method only removes the block from the config, and is called when you click the remove button in the layout drag drop page

Ah, garbage, I'm afraid, you are right... would have been too nice...

**Comment by indigoxela** [comment]
So a different concept is needed. No other method/hook has the information we need, at least as far as I can see. Or is there any?

The only thing that comes to my mind is to add the "whatever-job" to the Backdrop queue and if ... an hour later or so ... the block uuid still exists in that layout, cancel the "whatever-job", otherwise run it.
While still not perfect (when do unsaved changes expire?), there's a little more safety to prevent side effects.

Too weird? Worth a try?

**Comment by docwilmot** [comment]
Or perhaps add a `blocks_removed = array()` property to Layout thats populated on calling `removeBlock()` but cleared on cancelling. Then in `Layout::save()` invoke your new hook passing that array? It would need to document a caveat that certain static blocks (like maybe the `Powered by Backdrop` block) could be removed but re-added before `save()` is called, but it would be sufficient for most cases I'd think.

**Comment by indigoxela** [comment]
> a blocks_removed = array() property to Layout thats populated on calling removeBlock() but cleared on cancelling

That sounds promising. Currently layouts simply do a unset() on data parts (blocks). So recording changes, or possibly some kind of "diff" could help a lot.

**Comment by indigoxela** [comment]
Another day, another PR :wink:  https://github.com/backdrop/backdrop/pull/3019

@docwilmot - is this roughly what you had in mind?

There's already a hook in Layout::save, I don't think we need another one.
https://api.backdropcms.org/api/backdrop/core%21modules%21layout%21layout.api.php/function/hook_layout_presave/1

It seems to do what it should. Feedback is welcome.

**Comment by docwilmot** [comment]
Quick look only and it looks reasonable. But I'd call the variable blocks_removed instead to match the button action. And need some cleanup of the property value for when the user cancels instead of save, or re-added the block after clicking remove.

**Comment by indigoxela** [comment]
> I'd call the variable blocks_removed instead to match the button action

Good point, "deleted" is misleading here.

> need some cleanup of the property value for when the user cancels

Are you sure? Doesn't layout_settings_form_reset() already take care of that?

> ...or re-added the block after clicking remove

Yep, reusable block types need some attention, probably in function addBlock(). The PR is incomplete, of course.

**Comment by docwilmot** [comment]
youre quite right

**Comment by indigoxela** [comment]
Regarding the filtering of re-added blocks... I'm not so sure anymore, if we need to take care of it. 

Let me explain:

A block in a layout is identifiable only by its uuid. As soon as a block gets removed (button Remove, function removeBlock), this uuid gets dropped.

If a block of the same type (plugin) gets added again after removal (but before saving the layout), a new uuid is created, hence its not the same block. It's a new instance of the same plugin, it possibly even has different settings. And there can be several instances of the same plugin in a layout.

If one cancels layout changes (button Cancel), blocks keep their uuid's, but in this case the hook called from Layout::save doesn't run anyway.

Or are there other opinions?

**Comment by herbdool** [comment]
That makes sense to me. So what's the best way to test this PR?

**Comment by indigoxela** [comment]
How to test it:
* Add a hero block (or use an existing one)
* Add an image to the _Block content_
* Save that block
* Save that block again...
* Inspect file usage at /admin/content/files in parallel

**Comment by indigoxela** [comment]
How to test:
* Add a hero block to a layout
* Add an image to the Background image field of that hero block
* Save block and layout
* Inspect file usage
* Remove that block from that layout, save layout
* Inspect file usage

To test that file usage isn't touched if the layout wasn't saved, cancel the layout after you clicked the remove button.

Please note: there are other issues with file usage in layouts/blocks. Currently the hook is only implemented for BlockHero and only for the "Background image" field.

**Comment by herbdool** [comment]
@indigoxela okay that part works. I added hero image with background image, saved. Then removed the block and saved. It goes from permanent to temporary.

The thing that didn't work is if I first remove the background image but leave the block and save the layout. The image stays marked as permanent.

**Comment by indigoxela** [comment]
>  if I first remove the background image but leave the block and save the layout.

Good catch... `$block->settings['image']` is empty then. It's no block removal then, but a slightly different issue, maybe covered by https://github.com/backdrop/backdrop-issues/issues/4215.

**Comment by indigoxela** [comment]
@herbdool just in case you tested directly on my PR - it lags behind a bit, hence doesn't contain yet the fix from #4215 (which has been merged after my PR).

If you also want to test the behavior when _only_ the image, _not_ the block has been removed, or combinations of it, you probably need to [apply a patch](https://patch-diff.githubusercontent.com/raw/backdrop/backdrop/pull/3019.patch) to a more recent install.

**Comment by herbdool** [comment]
Would you be able to update your PR from upstream? Or I wonder if restarting the sandbox will apply the PR against the latest commits.

**Comment by herbdool** [comment]
Never mind. I tested a recent sandbox and it works there.

**Comment by herbdool** [comment]
I think it looks good. At least it will work for hero blocks now. And perhaps there'll need to be a follow-up for other block types with images.

**Comment by BWPanda** [comment]
@indigoxela I've left a code review requesting some small changes based on coding standards. Otherwise looks good to me and works as detailed above.

**Comment by indigoxela** [comment]
@BWPanda many thanks for your review, the PR has been updated accordingly.

**Comment by BWPanda** [comment]
Thanks @indigoxela, looks good! RTBC from me.

**Comment by BWPanda** [comment]
I've merged https://github.com/backdrop/backdrop/pull/3019 (my first one!) into 1.x. Thanks @indigoxela for the suggestion and PR, and to @docwilmot and @herbdool for discussion/feedback!

**Comment by klonos** [comment]
@BWPanda I've created a follow-up PR to add some comments in the code: https://github.com/backdrop/backdrop/pull/3097 (bases on the discussions between @indigoxela @docwilmot and @herbdool)

**Comment by BWPanda** [comment]
Thanks @klonos, @indigoxela, @docwilmot & @herbdool. Merged https://github.com/backdrop/backdrop/pull/3097 into 1.x.
